utils/wasm-gen/src/config/syscalls/param.rs

```diff
@@ -0,0 +1,169 @@
+// This file is part of Gear.
+
+// Copyright (C) 2021-2023 Gear Technologies Inc.
+// SPDX-License-Identifier: GPL-3.0-or-later WITH Classpath-exception-2.0
+
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+//! Entities describing sys-call param, more precisely, it's allowed values.
+//!
+//! Types here are used to create [`crate::SysCallsConfig`].
+
+use arbitrary::{Result, Unstructured};
+use gear_wasm_instrument::{
+    parity_wasm::elements::ValueType,
+    syscalls::{ParamType, SysCallName},
+};
+use std::{collections::BTreeMap, ops::RangeInclusive};
+
+#[derive(Debug)]
+pub(crate) enum ProcessedSysCallParams {
+    Alloc,
+    Value {
+        value_type: ValueType,
+        allowed_values: Option<SysCallParamAllowedValues>,
+    },
+    MemoryArray,
+    MemoryPtrValue,
+}
+
+pub(crate) fn process_sys_call_params(
+    name: Option<SysCallName>,
+    params: &[ParamType],
+    params_config: &SysCallsParamsConfig,
+) -> Vec<ProcessedSysCallParams> {
+    let skip_mem_array = name
+        .map(|name| matches!(name, SysCallName::SendInput | SysCallName::SendInputWGas))
+        .unwrap_or_default();
+
```

Comment:
If I understand right code in this function main complexity of this method is to identifies whether the param is simple ptr or ptr to array.
May be better to make im ParamType:
```
/// Syscall param type.
#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord)]
pub enum ParamType {
    Size,            // i32 buffers size in memory
    Ptr(Option<usize>),             // i32 pointer
    Gas,             // i64 gas amount
    MessagePosition, // i32 message position
    Duration,        // i32 duration in blocks
    Delay,           // i32 delay in blocks
    Handler,         // i32 handler number
    Alloc,           // i32 alloc pages
    Free,            // i32 free page
}
```
Where `Option<usize>` is index of argument which is size of array